plutus-contract/src/Language/Plutus/Contract/Trace.hs

```diff
@@ -468,15 +484,32 @@ handleBlockchainEvents
        )
     => Wallet
     -> ContractTrace s e m a ()
-handleBlockchainEvents wallet = do
-    hks <- getHooks wallet
-    if waitingForBlockchainActions hks
-        then do
-            submitUnbalancedTxns wallet
-            handleUtxoQueries wallet
-            handleOwnPubKeyQueries wallet
-            handleBlockchainEvents wallet
-        else pure ()
+handleBlockchainEvents = handleBlockchainEventsTimeout defaultMaxIterations
+
+-- | Handle all blockchain events for the wallet, throwing an error
+--   if the given number of iterations is exceeded
+handleBlockchainEventsTimeout
+    :: ( MonadEmulator (TraceError e) m
+    , HasUtxoAt s
+    , HasWatchAddress s
+    , HasWriteTx s
+    , HasOwnPubKey s
+    , HasTxConfirmation s
+    )
+    => MaxIterations
+    -> Wallet
+    -> ContractTrace s e m a ()
+handleBlockchainEventsTimeout (MaxIterations i) wallet = go 0 where
+    go j | j >= i    = lift (throwError (HandleBlockchainEventsMaxIterationsExceeded (MaxIterations i)))
+         | otherwise = do
+            hks <- getHooks wallet
+            if waitingForBlockchainActions hks
+                then do
+                    submitUnbalancedTxns wallet
+                    handleUtxoQueries wallet
+                    handleOwnPubKeyQueries wallet
+                    go (j + 1)
+                else pure ()
```

Comment:
I don't know if this is a worth a`MultiWayIf`, but this might be neater if you lifted the `getHooks` call and collapsed the conditional...

``` haskell
foo = go 0
  where
    go j = do
      hks <- getHooks wallet
      if | j >= i -> lift (throwError (HandleBlockchainEventsMaxIterationsExceeded (MaxIterations i)))
         | waitingForBlockchainActions hks -> do
              submitUnbalancedTxns wallet
              handleUtxoQueries wallet
              handleOwnPubKeyQueries wallet
              go (j + 1)
         | otherwise -> pure ()
```